modules/internal/ChapelBase.chpl

```diff
@@ -1272,6 +1272,11 @@ module ChapelBase {
     return if x != nil then __primitive("dynamic_cast", t, casttmp) else __primitive("cast", t, nil);
   }
 
+  // this version handles borrow -> unmanaged
+  inline proc _cast(type t:unmanaged, x:borrowed) {
+    return if x != nil then __primitive("dynamic_cast", t, x) else __primitive("cast", t, nil);
```

Comment:
This overload implements a dynamic cast so would need a where clause. I expect that we actually need 2 new _cast functions. One is for the case in which the target type is a parent class of the source value's type. The other is when the target type is a child class / subclass of the source value's type.